modules/development/src/main/java/unnamed/mmo/server/dev/tool/DebugToolManager.java

```diff
@@ -0,0 +1,185 @@
+        if (tool == null || debounce(player)) return;
+
+        final ItemStack newItemStack = tool.leftClicked(player, itemStack, null, null);
+        if (itemStack != newItemStack) {
+            player.setItemInHand(event.getHand(), itemStack);
+        }
+    }
+
+
+    private void changeHeldSlot(@NotNull PlayerChangeHeldSlotEvent event) {
+        final Player player = event.getPlayer();
+        //todo pretty sure this needs to check if the swap happened in the held slot or not.
+
+        // Deselect old tool
+        final ItemStack fromItem = player.getInventory().getItemStack(player.getHeldSlot());
+        final DebugTool fromTool = REGISTRY.get(fromItem.getTag(DEBUG_TOOL_TAG));
+        if (fromTool != null) {
+            fromTool.exitedHand(player, fromItem);
+        }
+
+        final ItemStack toItem = player.getInventory().getItemStack(event.getSlot());
+        final DebugTool toTool = REGISTRY.get(toItem.getTag(DEBUG_TOOL_TAG));
+        if (toTool != null) {
+            toTool.enteredHand(player, toItem);
+        }
+    }
+
+
+    private void swapItem(@NotNull PlayerSwapItemEvent event) {
+        final Player player = event.getPlayer();
+
+        // Exit immediately if old-main hand item is not a debug tool
+        if (!event.getOffHandItem().hasTag(DEBUG_TOOL_TAG))
+            return;
+        event.setCancelled(true);
+
+        final Inventory toolSelector = new Inventory(InventoryType.CHEST_2_ROW, "Tool Selector");
+        toolSelector.addInventoryCondition(this::selectTool);
+        REGISTRY.values().stream()
+                .sorted(Comparator.comparing(DebugTool::name))
+                .forEach(tool -> toolSelector.addItemStack(tool.itemStack()
+                        .withTag(DEBUG_TOOL_TAG, tool.name())));
+        player.openInventory(toolSelector);
+    }
+
+    private void disconnect(@NotNull PlayerDisconnectEvent event) {
+        debounce.remove(event.getPlayer().getUuid());
+    }
+
+    private void selectTool(@NotNull Player player, int slot, @NotNull ClickType clickType, @NotNull InventoryConditionResult result) {
+        result.setCancel(true);
+        if (clickType != ClickType.LEFT_CLICK) return;
+
+        final Inventory inventory = player.getOpenInventory();
+        if (inventory == null) return;
+        final ItemStack clickedItem = inventory.getItemStack(slot);
+        if (clickedItem.isAir()) return;
+
+        final ItemStack oldItem = player.getInventory().getItemStack(player.getHeldSlot());
+        final DebugTool oldTool = REGISTRY.get(oldItem.getTag(DEBUG_TOOL_TAG));
+        if (oldTool != null) {
+            oldTool.exitedHand(player, oldItem);
+        }
+
+        player.getInventory().setItemStack(player.getHeldSlot(), clickedItem);
+        final DebugTool newTool = REGISTRY.get(clickedItem.getTag(DEBUG_TOOL_TAG));
+        newTool.enteredHand(player, clickedItem);
+
+        player.closeInventory();
+    }
+
+
+    private boolean debounce(@NotNull Player player) {
```

Comment:
I would recommend a javadoc comment here so people can easily tell what true/false means from this method